src/cli.rs

```diff
@@ -0,0 +1,153 @@
+use crate::{merkle_tree::MerkleTree, proof_element::ProofElement};
+use anyhow::{Error, Result};
+use clap::{Parser, Subcommand};
+use std::{
+    fs::File,
+    io::{self, BufRead, BufReader, Write},
+    path::PathBuf,
+};
+
+/// CLI tool for tree-related operations
+#[derive(Parser, Debug)]
+#[command(name = "tree")]
+#[command(about = "Tree command-line tool", long_about = None)]
+pub struct Cli {
+    #[command(subcommand)]
+    command: Commands,
+
+    /// Merkle tree instance to hold the state
+    #[arg(skip)]
+    mktree: MerkleTree,
+}
+
+#[derive(Subcommand, Debug)]
+enum Commands {
+    /// Shows the tree structure
+    Show,
+    /// Adds an element to the tree
+    Add { element: String },
+    /// Verifies a proof for a given hash
+    Verify { hash: String, proof_file: PathBuf },
+    /// Generates a proof for a given hash.
+    Proof { hash: String },
+    /// Builds a tree with the provided elements.
+    Build { elements: Vec<String> },
+}
+
+impl Default for Cli {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Cli {
+    pub fn new() -> Self {
+        Cli {
+            command: Commands::Show, // Placeholder, will be overridden by parsing
+            mktree: MerkleTree::new_empty(),
+        }
+    }
+
+    pub fn run(&mut self) -> Result<()> {
+        loop {
+            // Display prompt
+            print!("tree> ");
+            io::stdout().flush()?; // Flush prompt to the terminal
+
+            // Read input from the user
+            let mut input = String::new();
+            io::stdin().read_line(&mut input)?;
+
+            // Trim the input and split it into parts
+            let input = input.trim();
+            let parts: Vec<&str> = input.split_whitespace().collect();
+
+            // Exit condition
+            if input == "exit" {
+                break;
+            }
+
+            // Skip empty input
+            if parts.is_empty() {
+                continue;
+            }
+
+            // Parse command
+            let command = match Cli::try_parse_from(parts.iter()) {
+                Ok(cli) => cli.command,
+                Err(err) => {
+                    println!("{}", err);
+                    continue;
+                }
+            };
+
+            // Execute the command and handle any errors
+            if let Err(e) = self.execute_command(command) {
+                println!("Error: {}", e);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn execute_command(&mut self, command: Commands) -> Result<()> {
```

Comment:
Here I would do this:

```suggestion
    fn execute_command(&mut self, command: Commands) -> Result<(), MerkleError> {
```

Reply:
I agree. Done!